table new api

jburel opened the pull request:
# What this PR does

Remove usage of the old API

Table creation and manipulation tested with the following version of pytables:

* 3.3.0 (Debian9)
* 3.2.x (Ubuntu16.04, CentOS 7)
* 3.1.x (ubuntu 14.04)

# Testing this PR

check that the integration tests and training examples are green


# Related reading

https://trello.com/c/l2UuwAee/74-tables-failure-on-debian-9-pytables-330

--depends-on #5409

jburel committed:
Remove usage of old api

jburel committed:
Merge remote-tracking branch 'origin/develop' into tables

jburel committed:
Update method

jburel committed:
Allow to specify policy and handle error if strict

jburel committed:
Specify policy in test

jburel commented:
Changes tested in https://github.com/openmicroscopy/openmicroscopy/pull/5419
travis is now green

jburel committed:
Merge remote-tracking branch 'origin/develop' into tables

jburel commented:
Failure on eel with older version of tables
Minimum requirement to be discussed i.e. do we support the version installed by the package manager on a  given OS. 3.1 is the minimum of the "supported platform"? or do we still support older version e.g. 2.1 or higher
In that case, for each change in this PR we will need a fallback

manics commented on components/tools/OmeroPy/src/omero/tables.py:
```diff
@@ -229,14 +229,32 @@ def openfile(self, mode):
+                    "which allows\n"
+                    "files to be opened multiple times at once\n"
+                    .format(version=tables.__version__,
+                            hdf_version=tables.get_hdf5_version()))
```
Do you have a list of which versions of HDF5 and PyTables are required, and how does it compare with the current distribution versions? Also does this ever occur in practice- isn't `HDFLIST = HdfList()` meant to prevent OMERO.tables opening the same file multiple times?

jburel replied:
In previous versions of pytables, hdflist prevents you from opening the file but that was not prevented at the level of pytables when using openFile
``addOrThrow`` will always throw at open_file w/o strict and ignore 
```
fileno = hdffile.fileno()

        try:
            portalocker.lockno(
                fileno, portalocker.LOCK_NB | portalocker.LOCK_EX)
        except portalocker.LockException:
            hdffile.close()
            raise omero.LockTimeout(
                None, None,
                "Cannot acquire exclusive lock on: %s" % hdfpath, 0)
        except:
            hdffile.close()
            raise
```

manics commented on components/tools/OmeroPy/src/omero/tables.py:
```diff
@@ -217,9 +217,9 @@ def __init__(self, file_path, hdf5lock):
 
-    def openfile(self, mode):
+    def openfile(self, mode, policy='default'):
+        tables.file._FILE_OPEN_POLICY = policy
```
`tables.file._FILE_OPEN_POLICY` is a global module level variable. Do you know in what situations it isn't already set to `default`?

jburel replied:
in newer version of pytables it is set to ``strict``

manics commented on components/tools/OmeroPy/test/unit/tablestest/test_hdfstorage.py:
```diff
@@ -310,10 +310,11 @@ def testLocking(self, monkeypatch):
-                        rootUEP='/').AndReturn(open(tmp))
+        # HdfList uses portalocker, test by mocking tables.open_file
+        self.mox.StubOutWithMock(tables, 'open_file')
+        tables.file._FILE_OPEN_POLICY = 'default'
```
See previous comment about `tables.file._FILE_OPEN_POLICY`

manics commented:
List of Python files using tables:

```
$ git grep -Ei 'import.*tables' -- '*.py'

components/tools/OmeroPy/src/omero/all.py:    IceImport.load("omero_Tables_ice")
components/tools/OmeroPy/src/omero/columns.py:IceImport.load("omero_Tables_ice")
components/tools/OmeroPy/src/omero/columns.py:    tables = __import__("tables")  # Pytables
components/tools/OmeroPy/src/omero/grid/__init__.py:IceImport.load("omero_Tables_ice")
components/tools/OmeroPy/src/omero/install/bzip2_tool.py:    import tables
components/tools/OmeroPy/src/omero/install/perf_test.py:        import tables
components/tools/OmeroPy/src/omero/tables.py:tables = __import__("tables")  # Pytables
components/tools/OmeroPy/src/runTables.py:    import omero.tables
components/tools/OmeroPy/test/integration/tablestest/test_backwards_compatibility.py:            import tables
components/tools/OmeroPy/test/integration/tablestest/test_service.py:import omero.tables
components/tools/OmeroPy/test/integration/tablestest/test_service.py:            import tables
components/tools/OmeroPy/test/unit/tablestest/test_hdfstorage.py:import omero.tables
components/tools/OmeroPy/test/unit/tablestest/test_hdfstorage.py:import tables
components/tools/OmeroPy/test/unit/tablestest/test_servants.py:import omero.tables
```



For future PRs:

There's some old code in https://github.com/openmicroscopy/openmicroscopy/blob/v5.4.0-m2/components/tools/OmeroPy/src/omero/columns.py#L283 which should be cleaned up if we're bumping the minimum tables requirement

Can https://github.com/openmicroscopy/openmicroscopy/blob/v5.4.0-m2/components/tools/OmeroPy/src/omero/install/bzip2_tool.py be deleted?
```
$git grep bzip2_tool

components/tools/OmeroPy/src/omero/install/bzip2_tool.py:def bzip2_tool(disable=False):
components/tools/OmeroPy/src/omero/install/bzip2_tool.py:            bzip2_tool(disable=which)
components/tools/OmeroPy/src/omero/install/bzip2_tool.py:        print "bzip2_tool failed: ", e
```

joshmoore commented:
> Can https://github.com/openmicroscopy/openmicroscopy/blob/v5.4.0-m2/components/tools/OmeroPy/src/omero/install/bzip2_tool.py be deleted?

Likely unneeded since 2ec6a6dae12cc298d5b6675c346bd3a652391797

jburel commented:
Excluding for now 
due to test failure (older version of pytables in CI)

--exclude

joshmoore committed:
Support old and new PyTables in tests

joshmoore committed:
Refactor HdfStorage out of tables.py

In order to support old and new PyTables versions, this
allows for choosing during service creation the storage
provider. This is the first step in to allowing non-HDF
backends.

joshmoore committed:
Add logging of storage type

jburel committed:
Merge pull request #13 from joshmoore/tables-legacy

Tables legacy layer

jburel committed:
Fix flake8 issue

joshmoore commented:
Closing in favor of https://github.com/openmicroscopy/openmicroscopy/pull/5481
